### PR TITLE
feat(InstanceProfile): Add IAM InstanceProfile Managed Resource

### LINKED ISSUE #1064: Add InstanceProfile CRD
### What resource do you need?
AWS cli parameter is:
`create-instance-profile` and `add-role-to-instance-profile`

### What is your use case?

We are building VM which runs multiple commands against multiple endpoints where credentials are needed. Instead of passing these secrets directly into userdata field we can use `Instance Profile` to read credentials from `AWS SecretManager`.

### Would you be willing to contribute it using [code generator](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md)?
Sure, but might take some time at this point as we just started to use crossplane.

### PR DESCRIPTION
**Pull request by ytsarev**
### Description of your changes

This PR adds support for IAM InstanceProfile including the IAM role association.

Caveats:
* Dedicated API call was required for role association https://docs.aws.amazon.com/IAM/latest/APIReference/API_AddRoleToInstanceProfile.html, no chance to make it declaratively, aws-sdk/api does not allow it
* Thus, I had to implement this special request in `preCreate` hook
* Same goes for associated https://docs.aws.amazon.com/IAM/latest/APIReference/API_RemoveRoleFromInstanceProfile.html in the `preDelete` hook 
* Only single IAM role association at the moment 

Fixes #1064

Depends on #1122 to be merged first.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

e2e:
```
k apply -f examples/iam/roleforinstanceprofile.yaml  
k apply -f examples/iam/instanceprofile.yaml

k get instanceprofiles.iam.aws.crossplane.io                                                                                                    
NAME                  READY   SYNCED   EXTERNAL-NAME
someinstanceprofile   True    True     someinstanceprofile

aws iam get-instance-profile --instance-profile-name someinstanceprofile                                                                         ✔ 
{
    "InstanceProfile": {
        "Path": "/",
        "InstanceProfileName": "someinstanceprofile",
        "InstanceProfileId": "AIPAY4AFTTSETG45HA4XQ",
        "Arn": "arn:aws:iam::xxx:instance-profile/someinstanceprofile",
        "CreateDate": "2022-02-18T15:47:05Z",
        "Roles": [
            {
                "Path": "/",
                "RoleName": "somerole",
                "RoleId": "AROAY4AFTTSE2QO5XMMRU",
                "Arn": "arn:aws:iam::xxxx:role/somerole",
                "CreateDate": "2022-02-18T15:46:49Z",
                "AssumeRolePolicyDocument": {
                    "Version": "2012-10-17",
                    "Statement": [
                        {
                            "Effect": "Allow",
                            "Principal": {
                                "Service": [
                                    "s3.amazonaws.com",
                                    "lambda.amazonaws.com",
                                    "eks-fargate-pods.amazonaws.com",
                                    "ec2.amazonaws.com",
                                    "eks.amazonaws.com"
                                ]
                            },
                            "Action": "sts:AssumeRole"
                        }
                    ]
                }
            }
        ],
        "Tags": [
            {
                "Key": "k1",
                "Value": "v1"
            }
        ]
    }
}


```

### CODE REVIEW

**Comment by haarchri** [comment]
@ytsarev  can you rebase ? ;) sdk bump is done

**Comment by ytsarev** [comment]
@haarchri that's amazing, thanks! PR is rebased :)

File: pkg/controller/iam/instanceprofile/setup.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	svc := iam.New(sess)
+	input := &iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: aws.String(cr.Name),
+		RoleName:            cr.Spec.ForProvider.Role,
+	}
+
+	_, err = svc.AddRoleToInstanceProfile(input)
+	return cre, err
+}
```
Comment (haarchri) [inline pkg/controller/iam/instanceprofile/setup.go:102]:
```suggestion
type updater struct {
	client svcsdkapi.IAMAPI
}

func (u *updater) postCreate(ctx context.Context, cr *svcapitypes.InstanceProfile, resp *svcsdk.CreateInstanceProfileOutput, cre managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
	if err != nil {
		return managed.ExternalCreation{}, err
	}

	input := &svcsdk.AddRoleToInstanceProfileInput{
		InstanceProfileName: aws.String(meta.GetExternalName(cr)),
		RoleName:            cr.Spec.ForProvider.Role,
	}

	_, err = u.client.AddRoleToInstanceProfileWithContext(ctx, input)
	return cre, err
}
```

File: pkg/controller/iam/instanceprofile/setup.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	svc := iam.New(sess)
+	input := &iam.RemoveRoleFromInstanceProfileInput{
+		InstanceProfileName: aws.String(cr.Name),
+		RoleName:            cr.Spec.ForProvider.Role,
+	}
+
+	_, err = svc.RemoveRoleFromInstanceProfile(input)
+	return false, err
+}
```
Comment (haarchri) [inline pkg/controller/iam/instanceprofile/setup.go:118]:
```suggestion
func (u *updater) preDelete(ctx context.Context, cr *svcapitypes.InstanceProfile, obj *svcsdk.DeleteInstanceProfileInput) (bool, error) {
	obj.InstanceProfileName = aws.String(meta.GetExternalName(cr))
	input := &svcsdk.RemoveRoleFromInstanceProfileInput{
		InstanceProfileName: aws.String(meta.GetExternalName(cr)),
		RoleName:            cr.Spec.ForProvider.Role,
	}

	_, err := u.client.RemoveRoleFromInstanceProfileWithContext(ctx, input)
	return false, err
}
```

File: pkg/controller/iam/instanceprofile/setup.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	svcapitypes "github.com/crossplane/provider-aws/apis/iam/v1alpha1"
+	aws "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// SetupInstanceProfile adds a controller that reconciles InstanceProfile.
+func SetupInstanceProfile(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(svcapitypes.InstanceProfileGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.preObserve = preObserve
```
Comment (haarchri) [inline pkg/controller/iam/instanceprofile/setup.go:47]:
```suggestion
			u := &updater{client: e.client}
			e.preObserve = preObserve
```

File: pkg/controller/iam/instanceprofile/setup.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+*/
+
+package instanceprofile
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/iam"
```
Comment (haarchri) [inline pkg/controller/iam/instanceprofile/setup.go:24]:
```suggestion
```

File: pkg/controller/iam/instanceprofile/setup.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	svcsdk "github.com/aws/aws-sdk-go/service/iam"
```
Comment (haarchri) [inline pkg/controller/iam/instanceprofile/setup.go:25]:
```suggestion
	svcsdk "github.com/aws/aws-sdk-go/service/iam"
	svcsdkapi "github.com/aws/aws-sdk-go/service/iam/iamiface"
```

**Review by haarchri — COMMENTED** [review]
added a few code comments - think its better to use `svcsdkapi.IAMAPI` like in other provider-aws resource implementations and call `add/removeRoleFromInstanceProfileWithContext` what do you think @ytsarev 

in general the creation / deletion is working =)

**Review by haarchri — APPROVED** [review]
LGTM - thanks @ytsarev

**Pull request merged**